### PR TITLE
NumPy to Torch for lightning/fabric

### PR DESCRIPTION
**Pull request by ishandutta0098**
## What does this PR do?
This PR is a part of #16649 
I have removed the numpy requirement from the files within src/lightning/fabric including:
- `src/lightning/fabric/plugins/environments/mpi.py`
- `src/lightning/fabric/loggers/tensorboard.py`

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/fabric/loggers/tensorboard.py [hunk 1]
```diff
@@ -303,7 +303,7 @@ def _get_next_version(self) -> int:
     def _sanitize_params(params: Dict[str, Any]) -> Dict[str, Any]:
         params = _utils_sanitize_params(params)
         # logging of arrays with dimension > 1 is not supported, sanitize as string
-        return {k: str(v) if isinstance(v, (Tensor, np.ndarray)) and v.ndim > 1 else v for k, v in params.items()}
+        return {k: str(v) if isinstance(v, (Tensor, torch.Tensor)) and v.ndim > 1 else v for k, v in params.items()}
```
Comment (carmocca) [inline src/lightning/fabric/loggers/tensorboard.py:306]:
```suggestion
        return {k: str(v) if hasattr(v, "ndim") and v.ndim > 1 else v for k, v in params.items()}
```

File: src/lightning/fabric/plugins/environments/mpi.py [hunk 2]
```diff
@@ -109,8 +108,8 @@ def _init_comm_local(self) -> None:
         hostname = socket.gethostname()
         all_hostnames = self._comm_world.gather(hostname, root=0)
         # sort all the hostnames, and find unique ones
-        unique_hosts = np.unique(all_hostnames)
+        unique_hosts = sorted(list(set(all_hostnames)))
```
Comment (carmocca) [inline src/lightning/fabric/plugins/environments/mpi.py:111]:
```suggestion
        unique_hosts = sorted(set(all_hostnames))
```

**Review by carmocca — APPROVED** [review]

File: src/lightning/fabric/loggers/tensorboard.py [hunk 3]
```diff
@@ -303,7 +303,7 @@ def _get_next_version(self) -> int:
     def _sanitize_params(params: Dict[str, Any]) -> Dict[str, Any]:
         params = _utils_sanitize_params(params)
         # logging of arrays with dimension > 1 is not supported, sanitize as string
-        return {k: str(v) if isinstance(v, (Tensor, np.ndarray)) and v.ndim > 1 else v for k, v in params.items()}
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/fabric/loggers/tensorboard.py:306]: Do you think that my "hasattr" suggestion is not sufficient?

[the remark this answers is not in the document]
Comment (ishandutta0098) [reply src/lightning/fabric/loggers/tensorboard.py:306]: @awaelchli  Why do we need a numpy check in the `tensorboard.py` file?

[the remark this answers is not in the document]
Comment (ishandutta0098) [reply src/lightning/fabric/loggers/tensorboard.py:306]: Okay I get it, so for all the changes in this file I should ensure that both the options of numpy and torch are available by checking if numpy is available. 

Shall I implement a `numpy_is_available` function in `src/lightning/fabric/utilities/imports.py` or use the function available in `src/lightining/app/utilities/imports.py`

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/fabric/loggers/tensorboard.py:306]: Here we are lucky that both torch.Tensor and numpy.ndarray have the same attribute name for the information we need. So the trick that @carmocca found is sufficient. We can add the numpy_available check elsewhere if it is needed but it probably won't be.

**Review by awaelchli — APPROVED** [review]

**Pull request merged**